NitroxServer-Subnautica/Resources/Parsers/PrefabPlaceholderGroupsParser.cs

```diff
@@ -132,35 +195,70 @@ private ConcurrentDictionary<string, string[]> GetAllPrefabPlaceholdersGroupsFas
+                        {
+                            spawnRandomHash[i] = monoScript["m_PropertiesHash"][i].AsByte;
+                        }
+                        break;
+                    case "PrefabPlaceholdersGroup":
+                        prefabPlaceholdersGroupHash = new byte[16];
+                        for (int i = 0; i < 16; i++)
+                        {
+                            prefabPlaceholdersGroupHash[i] = monoScript["m_PropertiesHash"][i].AsByte;
+                        }
+                        break;
                 }
-
-                prefabPlaceholdersGroupHash = new byte[16];
-                for (int i = 0; i < 16; i++)
-                {
-                    prefabPlaceholdersGroupHash[i] = monoScript["m_PropertiesHash"][i].AsByte;
-                }
-
-                break;
             }
 
-            if (prefabPlaceholdersGroupHash.Length != 0)
+            if (prefabPlaceholdersGroupHash.Length > 0 && spawnRandomHash.Length > 0)
             {
                 break;
             }
         }
 
-        Parallel.ForEach(addressableCatalog.Skip(aaIndex), (keyValuePair) =>
+        // Now use the bundle paths and the hashes to find out which items from the catalog are important
+        // We fill prefabPlaceholdersGroupPaths and RandomPossibilitiesByClassId when we find objects with a SpawnRandom
+        Parallel.ForEach(addressableCatalog, (keyValuePair) =>
         {
+            string[] assetPaths = keyValuePair.Value;
+            
             AssetsBundleManager bundleManagerInst = am.Clone();
-            BundleFileInstance bundleFile = bundleManagerInst.LoadBundleFile(bundleManagerInst.CleanBundlePath(keyValuePair.Value[0]));
-            AssetsFileInstance assetFileInstance = bundleManagerInst.LoadAssetsFileFromBundle(bundleFile, 0);
+            AssetsFileInstance assetFileInstance = bundleManagerInst.LoadBundleWithDependencies(assetPaths);
 
-            if (assetFileInstance.file.Metadata.TypeTreeTypes.Any(typeTree => typeTree.TypeId == (int)AssetClassID.MonoBehaviour && typeTree.TypeHash.data.SequenceEqual(prefabPlaceholdersGroupHash)))
+            foreach (TypeTreeType typeTreeType in assetFileInstance.file.Metadata.TypeTreeTypes)
             {
-                prefabPlaceholdersGroupPaths.TryAdd(keyValuePair.Key, keyValuePair.Value);
+                if (typeTreeType.TypeId != (int)AssetClassID.MonoBehaviour)
+                {
+                    continue;
+                }
+
+                if (typeTreeType.TypeHash.data.SequenceEqual(prefabPlaceholdersGroupHash))
+                {
+                    prefabPlaceholdersGroupPaths.TryAdd(keyValuePair.Key, keyValuePair.Value);
+                    break;
+                }
+                else if (typeTreeType.TypeHash.data.SequenceEqual(spawnRandomHash))
+                {
+                    AssetsFileInstance assetFileInst = am.LoadBundleWithDependencies(assetPaths);
```

Comment:
Unless I'm really reading this wrong isn't this the same as the AssetsFileInstance before the for loop? Why can't we reuse that rather than reloading the same file?